src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems

```diff
@@ -1005,10 +1029,13 @@
     <Compile Include="$(MSBuildThisFileDirectory)Interop\Windows\User32\Interop.LoadString.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Interop\Windows\User32\Interop.SendMessageTimeout.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Microsoft\Win32\SafeHandles\SafeLibraryHandle.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\Environment.Win32.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)System\Environment.Win32.Registry.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\IO\FileStream.Win32.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\TimeZoneInfo.Win32.cs" />
   </ItemGroup>
   <ItemGroup Condition="$(TargetsWindows) and '$(EnableWinRT)' == 'true'">
+    <Compile Include="$(MSBuildThisFileDirectory)System\Environment.WinRT.cs" />
```

Comment:
@jkotas, I only had it here as a way to share the code currently in https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/shared/System/Environment.WinRT.cs.  But if it doesn't need to be shared, I can just leave it as corert-specific and delete it from here. Is there a problem doing that rather than doing the reflection/callback approach?

Reply:
`Windows.Foundation.Metadata` and `Windows.Storage` are WinRT namespaces that come from Windows.winmd. Using WinRT interfaces in CoreLib is:
- Pain in CoreLib. You have to create a local C# clone of the WinRT interface in CoreLib. (Look for `[WindowsRuntimeImport]`.)
- Forbidden in ProjectN. ProjectN toolchain does not support having WinRT interop in CoreLib. Environment.WinRT.cs won't work in ProjectN CoreLib.

The typical way people have solved this WinRT problem is:
- For CoreCLR, either clone the WinRT interface (if it is small or if it is required for other reasons); or use the reflection callback to CoreFX if the WinRT interface is larger.
- For ProjectN, neither cloning or reflection works. So folks ended up adding the calls to WinRT APIs to `System.Private.WinRTInterop.CoreLib` library that is hooked up as a callback via internal non-reflection based mechanism by the ProjectN toolchain. The `System.Private.WinRTInterop.CoreLib` has not been open sourced, changes to it need to be submitted to TFS directly. @MichalStrehovsky is the person to talk to about this.

Reply:
To preserve the current behavior, I think the implementation of `Environment.GetFolderPath` should:

- In CoreCLR: Call the WinRT implementation when `FEATURE_APPX` is defined in the build and `ApplicationModel.IsUap` is true.
- In ProjectN: Call the WinRT implementation when `ENABLE_WINRT` is define during the build and `WinRTInterop.UnsafeCallbacks.IsAppxModel()` is true.

CultureInfo has a good prior-art for this: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Globalization/CultureInfo.cs#L410 . I would make the `Environment.GetFolderPath` to follow same or very similar pattern.

I do not think it is important to make this pretty, or try hard to avoid duplication. 

> I get that I can't include this in coreclr

You could (in theory at least) if you choose to clone the WinRT interfaces in CoreLib.

Reply:
Ugh.  Ok.  If I'm understanding correctly, that means adding a type to corefx that corelib will depend on via reflection.  Are you ok with that temporarily not working as bits flow through the system, or would you prefer I add the support to corefx, get that checked in, wait until the binaries are consumed into coreclr, then move forward with this?

Reply:
> adding a type to corefx that corelib will depend on via reflection

I think that's fine. Tests in UAP legs may need to be temporarily disabled in the process, but I do not see a problem with it.